PRESIDECMS-1460 selectable export formbuilder

synmal opened the pull request:
This changes will allow users to choose which field to be included during exporting form builder submissions.

![Screenshot 2019-07-08 at 10 55 38](https://user-images.githubusercontent.com/44126659/60780068-9123e000-a16f-11e9-817e-1360772b959a.png)

This is done by adding a form field in formfield.xml
`<field name="export" control="yesnoswitch" default="true" sortorder="60" />`

Into the backend side, in the FormBuilderService.cfc, I added a condition which checks the formitem configuration if it is exportable or not (Refer to the changes in FormBuilderService.cfc). It will skip the formitem from being appended in the itemsToRender array if the export value is 0 or false.

maljer committed:
[twgit] Init feature 'feature-PRESIDECMS-1460_selectable_export_formbuilder'.

maljer committed:
Added export field in yesno switch in field config form

maljer committed:
Moved the field to formbuilder_formitem object instead.

maljer committed:
Revert "Moved the field to formbuilder_formitem object instead."
export value will still be in configuration struct anyways

This reverts commit 110924a65cf0b37ccf80eefa489689fc6d98761b.

maljer committed:
Admin now can choose which field to export during configuring individual formitem.

maljer committed:
Include in export option is true by default

DominicWatson commented on system/services/formbuilder/FormBuilderService.cfc:
```diff
@@ -911,8 +911,9 @@ component {
 		for( var i=1; i <= formItems.len(); i++ ) {
 			if ( formItems[i].type.isFormField ) {
 				var columns = renderingService.getItemTypeExportColumns( formItems[i].type.id, formItems[i].configuration );
+				var export  = formItems[i].configuration.export ?: 0 ;
```
I suggest this should be more defensive. The following approach is solid for ensuring boolean values:

```
var export = IsBoolean( formItems[ i ].configuration.export ?: "" ) && formItems[ i ].configuration.export;
```

DominicWatson replied:
Also, only get the columns if `export = true`, so switch lines 913 + 914 around and do:

```
var columns = export ? renderingService.getItemTypeExportColumns( formItems[i].type.id, formItems[i].configuration ) : [];
```

synmal replied:
The changes requested had been pushed.

DominicWatson reviewed (changes requested):
Hey @synmal - this looks great. Just a tiny bit of feedback to tweak. Thanks :)

maljer committed:
Added amendments as suggested by Dom.
In formfield.xml, the default="true" was remove due to settings does not persist when navigate out of the form config page.

synmal commented on system/forms/formbuilder/item-types/formfield.xml:
```diff
@@ -18,6 +18,7 @@ it will be merged with this form definition.
 			<field name="layout"    control="formbuilderFieldLayoutPicker" required="true"  sortorder="30" />
 			<field name="mandatory" control="yesnoswitch"                  required="false" sortorder="40" />
 			<field name="help"      control="textarea"                     required="false" sortorder="50" />
+			<field name="export"    control="yesnoswitch"                                   sortorder="60" />
```
default="true" was removed due to the front end will always show "true" even though it was set to false. This will happen when user navigate out from the config page and reloading. Removing the default will solve the problem.

DominicWatson commented on system/services/formbuilder/FormBuilderService.cfc:
```diff
@@ -910,8 +910,8 @@ component {
 			if ( formItems[i].type.isFormField ) {
-				var columns = renderingService.getItemTypeExportColumns( formItems[i].type.id, formItems[i].configuration );
-				var export  = formItems[i].configuration.export ?: 0 ;
+				var export  = isBoolean( formItems[i].configuration.export ?: "" ) && formItems[i].configuration.export ?: "";
```
I'm going to be really picky :) 

```
IsBoolean( somevar ?: "" ) && somevar;
// NOT:
IsBoolean( somevar ?: "" ) && somevar ?: "";
```

If `someVar ?: ""` is Boolean, then we know that it definitely exists so we don't need to the elvis operator a second time.

synmal replied:
Hi @DominicWatson, I removed the extra elvis operator. Thanks for the heads up.

DominicWatson reviewed (commented):
I'm being very picky :) Have commented again. Love this PR btw, just being a git!

maljer committed:
Merge tag 'v10.10.85' into feature-PRESIDECMS-1460_selectable_export_formbuilder

[twgit] Hotfix finish: hotfix-10.10.85

maljer committed:
Made amendments

DominicWatson commented:
Have finally got round to reviewing again, thanks @synmal . Merged into upcoming 10.11.0 release :)
